### ISSUE TITLE
Lag spike when using "Provider.clients" to get players.

### ISSUE THREAD

**Issue by megoesrawr** [SmartlyDressedGames/Unturned-3.x-Community#1288]
When using "Provider.clients" to get online players, it causes a huge lag spike and the whole server freezes for a second. 

EX:
`var players = Provider.clients.Where(k => UnturnedPlayer.FromSteamPlayer(k).HasPermission("permission"));
                foreach (var playr in players)
                {
                  dostuff();
                }`

**Comment by PandahutMiku** [comment]
The issue is probably with your code and not with Provider.Clients. Whatever your running is probably on the main thread and blocking it, freezing the server.

Provider Clients is used by a bunch of internal stuff:
![image](https://user-images.githubusercontent.com/17209937/70383777-623db000-1941-11ea-8ba5-aa062319f2b1.png)
If just using it was causing this, there would be constant freezes.

**Comment by Trojaner** [comment]
```
var players = Provider.clients.Where(k => UnturnedPlayer.FromSteamPlayer(k).HasPermission("permission")); foreach (var playr in players) { dostuff(); }
````

UnturnedPlayer.FromSteamPlayer has already a complexity of O(n) since if runs foreach on all players. You also do foreach for all players which results in a complexity of O(n²) (where n is the player count). 

In short, you are doing something like this:
```cs
foreach(var player in Provider.clients)
{
   // this is literally how FromSteamPlayer works
   UnturnedPlayer uplayer;
   foreach(var player2 in Provider.clients) 
   {
       if(player2.id == player.id)
       { 
             uplayer = new UnturnedPlayer(player2);
       }
   }

   doStuff();
}
```
O(n²) means, for 40 players there would be 40 * 40 = 1600 iterations!

To avoid this double foreach, maintain your own list of UnturnedPlayers every time a player joins and leaves. This is a RocketMod specific issue since it does not use the steamplayer your assign, but instead gets a new steamplayer from the id.

There is no problem on Unturneds side. Provider.clients is a List<SteamPlayer> with around 0-40  entries on average. Iterating such a list should be barely measurable.

In short; get rid of FromSteamPlayer and check if your "doStuff()" method runs heavy stuff like networking, downloading etc. on main thread.

/closerequest
